### ISSUE TITLE
Translate transform not being applied

### ISSUE THREAD

**Issue by mojavelinux** [mogest/prawn-svg#96]
It appears that the translate transform is not being applied anymore. I've provided the following project for demonstration. If you try it with prawn-svg 0.26.0 and 0.27.0, you'll see that you get different results. The image should appear at the top of the content area. Instead, it's written in the middle of the page. The difference seems to be the translation applied to the parent group.

**Comment by mojavelinux** [comment]
See https://github.com/mojavelinux/prawn-svg-issue-96

**Comment by mojavelinux** [comment]
Actually, it appears like the problem is that the image is vertically centered in the page...though I can't explain why.

**Comment by mojavelinux** [comment]
This appears to be a simple problem in the calculation of the output sizing.

If I add the following line in the draw method on Interface:

```
p [@document.sizing.output_width, @document.sizing.output_height]
```

With prawn-svg 0.26.0, I get:

```
[270.0, 111.3]
```

With prawn-svg 0.27.0, I get:

```
[270.0, 720.0]
```

I'll update the title to reflect the real problem.

**Comment by mojavelinux** [comment]
If I set the height explicitly, it all works correctly. So this is when either the height or width is dynamic.

**Comment by mojavelinux** [comment]
The unspecified dimension is growing to the available space instead of being scaled proportionally to the specified dimension.

**Comment by mojavelinux** [comment]
And specifically this is when the image document does not have a height and width (only a viewBox).

**Comment by mojavelinux** [comment]
It turns out, this change was intentional as a fix for #92. I strongly disagree with this change. The problem is that if a width is specified and not a height, we should still respect the viewBox. As it's implemented now, the SVG is blowing way past the viewBox to consume all the available height on the page. That's not how I would interpret using 100% and that is not how browsers work either.

**Comment by mojavelinux** [comment]
Here's a clearer example. The following two statements do not place the SVG at the same spot:

No viewBox:

```ruby
svg '<svg><rect width="200" height="200" fill="red"></rect></svg>',
  at: [0, bounds.top],
  width: 200
```

With viewBox:

```ruby
svg '<svg viewBox="0 0 200 200"><rect width="200" height="200" fill="red"></rect></svg>',
  at: [0, bounds.top],
  width: 200
```

By setting the viewBox, the calculation of the output width and height gets thrown off.

**Comment by mojavelinux** [comment]
What should have happened with the fix for #92 is that special logic should have been isolated to when the viewBox is not specified. Instead, it affected the case when a viewBox is specified, which was working properly in all cases.

**Comment by mojavelinux** [comment]
I hope that provides enough insight ;)

**Comment by mogest** [comment]
Thanks @mojavelinux, I agree it doesn't seem right. Can you review my commit, please?

**Comment by mojavelinux** [comment]
Almost there! There's still one case that doesn't seem right. Consider the following:

```ruby
svg '<svg viewBox="0 0 200 200"><rect width="200" height="200" fill="red"></rect></svg>',
  at: [0, bounds.top],
  width: bounds.width
text '^ ^ ^', align: :center
```

That renders as we expect...at the top of the page with width equal to height equal to the page content width.

Now take away the requested width:


```ruby
svg '<svg viewBox="0 0 200 200"><rect width="200" height="200" fill="red"></rect></svg>',
  at: [0, bounds.top]
text '^ ^ ^', align: :center
```

We expect the same results as the first. However, in this case, the box is drawn in the vertical center of the page. That's because the height of the SVG is expanding to fill the content area height...not taking into account the viewBox constraint.

**Comment by mogest** [comment]
That's consistent with how browsers render, though.  I can see the logic about restraining to the ratio of the viewBox when a width or height is explicitly provided, but when it's not, the SVG spec clearly says it defaults to 100% of the available space.

**Comment by mojavelinux** [comment]
I disagree that is how browsers render. I've added an example to prove it. See https://github.com/mojavelinux/prawn-svg-issue-96/blob/master/issue-96-box.html

I acknowledge the spec says to expand to the available space when no explicit width and height are given. But it still needs to respect the viewBox in doing so (like a balloon). In our example, the width expands to 100%, but the height needs to expand proportionally to respect the viewBox.

If we take a step back, this is actually pretty logical. If I set width to bounds.width, which is equivalent to 100%. Currently, the viewBox is respected in that case. If I don't set the width, it should expand to 100%, which is bounds.width. So those two scenarios are effectively the same. Setting width to bounds.width should be the same as not setting it. Instead, what we get is a breach of the viewBox in the latter case.

**Comment by mojavelinux** [comment]
Very interesting to note is that if you take away the viewBox in the issue-96-box.html example, what you discover is that the behavior of prawn-svg 0.26.0 is actually perfectly consistent with what the browsers do. The rectangle is not a square and it does not expand to fill the bounds. I don't know if that's really right. Just an observation. But I'll be honest that I don't know what an SVG is that doesn't have a viewBox or explicit dimensions. It seems amorphous to me. I'm really not concerned about that case because I consider that an invalid SVG.

**Comment by mojavelinux** [comment]
I guess you could make the case that what the browsers are doing here is wrong. What is currently implemented in master is only emulated in the browser using:

```xml
<svg viewBox="0 0 200 200" width="100%" height="100%">
<rect width="200" height="200" fill="red"></rect>
</svg>
```

There's clearly a judgment call to be made here. But I think it's more logical for the user if the SVG "canvas" only expands all the way to 100% if the viewBox permits it or if the viewBox is not there.

**Comment by mogest** [comment]
If you make an SVG document and display it in the browser:

```svg
<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg xmlns="http://www.w3.org/2000/svg" version="1.1" viewBox="0 0 200 200"><rect width="200" height="200" fill="red"></rect></svg>
```

Then it displays as if width was 100% and height was 100%, and that's what I'm emulating.  As you say, it doesn't make a whole lot of sense to not specify any width or height, and I think sticking with the SVG spec (i.e. unspecified == 100%) is the way to go because of that.

**Comment by hho** [comment]
I'd also love to see this fixed. Prawn is for PDFs and not a browser, so using 100% for unspecified is not helpful, especially for `height`.

**Comment by mojavelinux** [comment]
I stand firm that this is still a problem. When an explicit width is not specified, an unexplained gap appears above and below the image. And this surprises users every time. The workaround is to specify an explicit width, but it incurs a high support cost to have to continuously remind users of this.

**Comment by mojavelinux** [comment]
Here's another example: https://github.com/asciidoctor/asciidoctor-pdf/issues/886

**Comment by mojavelinux** [comment]
Here's the issue I have with the logic:

> the height of the SVG is expanding to fill the content area height...not taking into account the viewBox constraint.

The viewBox constraint should be honored.

**Comment by mojavelinux** [comment]
@mogest Thanks so much for resolving this. This fixes numerous rendering issues in Asciidoctor PDF. Now, SVGs generated by various diagramming tools are showing up where users expect them to show up. Happy campers. \o/
